ui/app/src/Views/GroupBar/Match.elm

```diff
@@ -0,0 +1,150 @@
+jaro s1 s2 =
+    if s1 == s2 then
+        1.0
+    else
+        let
+            l1 =
+                String.length s1
+
+            l2 =
+                String.length s2
+
+            z2 =
+                zip (List.range 1 l2) (String.toList s2)
+                    |> List.map (Tuple.mapSecond Char.toCode)
+
+            searchLength =
+                -- A character must be within searchLength spaces of the
+                -- character we are matching against in order to be considered
+                -- a match.
+                -- (//) is integer division, which removes the need to floor
+                -- the result.
+                ((max l1 l2) // 2) - 1
+
+            m =
+                zip (List.range 1 l1) (String.toList s1)
+                    |> List.map (Tuple.mapSecond Char.toCode)
+                    |> List.map (charMatch searchLength z2)
+                    |> List.foldl (++) []
+
+            ml =
+                List.length m
+
+            t =
+                m
+                    |> List.map (transposition z2 >> toFloat >> ((*) 0.5))
+                    |> List.sum
+
+            ml1 =
+                toFloat ml / toFloat l1
+
+            ml2 =
+                toFloat ml / toFloat l2
+
+            mtm =
+                (toFloat ml - t) / toFloat ml
+        in
+            if ml == 0 then
+                0
+            else
+                (1 / 3) * (ml1 + ml2 + mtm)
+
+
+winkler : String -> String -> Float -> Float
+winkler s1 s2 jaro =
+    if s1 == "" || s2 == "" then
+        0.0
+    else if s1 == s2 then
+        1.0
+    else
+        let
+            l =
+                commonPrefix s1 s2
+                    |> String.length
+                    |> toFloat
+
+            p =
+                0.1
+        in
+            jaro + ((l * p) * (1.0 - jaro))
+
+
+jaroWinkler : String -> String -> Float
+jaroWinkler s1 s2 =
+    if s1 == "" || s2 == "" then
+        0.0
+    else if s1 == s2 then
+        1.0
+    else
+        jaro s1 s2
+            |> winkler s1 s2
+
+
+commonPrefix : String -> String -> String
+commonPrefix s1 s2 =
+    if s1 == "" || s2 == "" then
+        ""
+    else if s1 == s2 then
+        String.left 4 s1
+    else
+        cp (String.toList s1) (String.toList s2) []
+            |> String.fromList
+
+
+cp : List Char -> List Char -> List Char -> List Char
+cp l1 l2 acc =
+    if List.length acc == 4 then
+        acc
+    else
+        case l1 of
+            [] ->
+                acc
+
+            x :: xs ->
+                case l2 of
+                    [] ->
+                        acc
+
+                    y :: ys ->
+                        if x == y then
+                            x :: cp xs ys acc
+                        else
+                            acc
+
+
+charMatch : Int -> List ( Int, Int ) -> ( Int, Int ) -> List ( Int, Int )
+charMatch matchRange list ( p, q ) =
+    -- TODO(w0rm): Is there a way to define this so it's not so strictly bound
```

Comment:
It might not be important, but being able to use a generic comparable number would be good. I think I tried using number initially, but then it needed to be comparable, and I didn't know how to say `charMatch : <number, comparable>` etc etc